pin.go

```diff
@@ -57,16 +62,45 @@ func (api *PinAPI) Ls(ctx context.Context, opts ...caopts.PinLsOption) ([]iface.
 		return nil, err
 	}
 
-	pins := make([]iface.Pin, 0, len(out.Keys))
-	for hash, p := range out.Keys {
-		c, err := cid.Parse(hash)
-		if err != nil {
-			return nil, err
+	pins := make(chan iface.Pin)
+	go func(ch chan<- iface.Pin) {
+		defer close(ch)
+		for hash, p := range out.Keys {
+			c, e := cid.Parse(hash)
+			if e != nil {
+				ch <- &pin{typ: p.Type, err: e}
+				return
+			}
+			ch <- &pin{typ: p.Type, path: path.IpldPath(c), err: e}
+		}
+	}(pins)
+	return pins, nil
+}
+
+// IsPinned returns whether or not the given cid is pinned
+// and an explanation of why its pinned
+func (api *PinAPI) IsPinned(ctx context.Context, p path.Path, opts ...caopts.PinIsPinnedOption) (string, bool, error) {
+	options, err := caopts.PinIsPinnedOptions(opts...)
+	if err != nil {
+		return "", false, err
+	}
+	var out pinRefKeyList
+	err = api.core().Request("pin/ls").
+		Option("type", options.WithType).
+		Option("arg", p.String()).
+		Exec(ctx, &out)
+	if err != nil {
+		//need an err type,or a constant value to compare
+		if strings.Index(err.Error(), "is not pinned") != -1 {
+			return "", false, nil
 		}
-		pins = append(pins, &pin{typ: p.Type, path: path.IpldPath(c)})
+		return "", false, err
 	}
 
-	return pins, nil
+	for _, obj := range out.Keys {
+		return obj.Type, true, nil
+	}
+	return "", false, errors.New("pin path not found")
```

Comment:
If the path is not found, wouldn't that be caught above where you check for error "is not pinned"? If so, this line here should be something like panic("unreachable").

Reply:
yes,this is generally impossible to reach.

Reply:
Thanks for addressing this. We discussed this error handling a bit further with @aschmahmann. We decided that this should not be a panic, because the last line can be reached also if this client is talking to a server with a mismatched version. I also opened an issue for providing a better way for handling the error checks above (which use brittle sub-string matching).

As a result, I've added the necessary fixes to this PR in https://github.com/ipfs/go-ipfs-http-client/pull/117. And I'll take care of seeing it through to merging.

Reply:
I am closing this PR, as it is now subsumed by https://github.com/ipfs/go-ipfs-http-client/pull/117.